### ISSUE TITLE
Interfacing 16X2 LCD   with BeagleBone Black Wireless

### ISSUE THREAD

**Issue by mikeshams** [chandrashekar1994/captson-project-summer-2020#6]

**Comment by manharjotkaur** [comment]
//16X2 LCD Display Interfacing with Baeglebone Black
//Using I2C bus....

#include <linux/i2c-dev.h> //i2c  Linux Library..
#include <stdio.h> //standard I/O Library..
#include <stdlib.h> //std. Library.
#include <unistd.h>
#include <fcntl.h> //Function Cntrol Lib.

#define I2C_BUS        "/dev/i2c-2" // I2C line on BBB..
#define I2C_ADDR       0x27        // Address of LCD module on i2c port.
 
int i2cFile;
// Check File opening Error as well as initializing....
void i2c_start() //I2C start function

{    if((i2cFile = open(I2C_BUS, O_RDWR)) < 0)
{     printf("Error: Failed to open I2C bus \n");
      exit(-1);}   
      if (ioctl(i2cFile, I2C_SLAVE, I2C_ADDR) < 0)
      { printf("Error: Failed to set I2C address.\n");
      exit(-1);}
}

void i2c_stop()
   {
    close(i2cFile);// closed file..
    }
void i2c_send_byte(unsigned char data) //Transfer data to i2c slave
{
   unsigned char byte[1];
   byte[0] = data;
   write(i2cFile, byte, sizeof(byte)); // writing info to i2c line
   usleep(1000);
}
void lcd_init()
{
   i2c_send_byte(0b00110100); // Data line set D7=0, D6=0, D5=1, D4=1, RS,RW=0 EN=1
   i2c_send_byte(0b00110000); // D7=0, D6=0, D5=1, D4=1, RS,RW=0 EN=0
   usleep(5000);              // delay 5.0 msec
   i2c_send_byte(0b00110100); //
   i2c_send_byte(0b00110000); // 
   usleep(100);               // delay 1msec
   i2c_send_byte(0b00110100); //
   i2c_send_byte(0b00110000); 
   usleep(5000);              // delay 5.0 msec
   i2c_send_byte(0b00100100); //
   i2c_send_byte(0b00100000); // set to 4-bit mode
   usleep(40);                // delay 4.0 usec
   LCD_cmd(0x28); // 4-bit mode with 2 lines char 5x8
   usleep(40);                // delay 0.4sec
   LCD_cmd(0x08);            //Display shutdown, cursor off
   usleep(40);     // delay 40usec
   LCD_cmd(0x08); //Display shutdown, cursor off
   LCD_cmd(0x01);
   usleep(40);   // delay 40usec
   LCD_cmd(0x06); //Increment cursor (shift cursor to right)
   usleep(40);                // delay 0.4msec
   LCD_cmd(0x0E); // Display on, cursor blinking
   LCD_cmd(0x80);   //Force cursor to set on starting ( 1st line)
 usleep(40);      // delay 0.4msec
   }

// function to give data to command register of LCD ....................
void
LCD_cmd (unsigned char cmd) 
{
char upper = cmd>>4 ;    //  msb of the data
char up1 =(upper<<4)|0x4; // addding enable pin high, rs pin 0, rw pin 0
char up2 =(upper<<4)|0x0; // addding enable pin low, rs pin 0, rw pin 0
char lower= cmd & 0x0f; // seperate lsb of the data
char lw1= (lower<<4) | 0x4; // addding enable pin high, rs pin 0, rw pin 0
char lw2= (lower<<4) | 0x0; // addding enable pin low, rs pin 0, rw pin 0
i2c_send_byte(up1); // sending data part by part
i2c_send_byte(up2);
i2c_send_byte(lw1);
i2c_send_byte(lw2);
  usleep (10);
}
void
LCD_data (unsigned char data) // function to give data to data register of LCD
{
char upper = data>>4 ; // seperate msb of the data
char up1 =(upper<<4)|0xd; // set enable pin high, rs pin 1, rw pin 0
char up2 =(upper<<4)|0x9;
char lower= data & 0x0f;  // seperate lsb of the data
char lw1= (lower<<4) | 0xd;  // set enable pin high, rs pin 1, rw pin 0
char lw2= (lower<<4) | 0x9; 

// sending data part by part..............

i2c_send_byte(up1); 
i2c_send_byte(up2);
i2c_send_byte(lw1);
i2c_send_byte(lw2);
  usleep (10);
}

// sending character one by one to bus...........

void
LCD_string (const char *buf) // send string of data
{
  while (*buf)
    {
      usleep (10); // 10us delay
      LCD_data (*buf); 
      buf++;
      usleep (10); // 10us delay
    }
}
void main() 
{
   i2c_start();      //initalizing i2c connection
   usleep(15000);             // wait 15msec
 lcd_init();       //initializing lcd display
 LCD_string(" SMART LIBRARY ");
 sleep(5);
 LCD_cmd(0x01);// clear display
 usleep(1000);
 LCD_cmd(0x80);//Force cursor to beginning ( 1st line)
 usleep(100);
 LCD_string (" PLEASE SCAN.. ");
 usleep(1000);
 LCD_cmd(0xC0);//Force cursor to beginning ( 2nd line)
 usleep(100);
 LCD_string (" RFID TAGS.. ");
 sleep(5);
LCD_cmd(0x01);// clear display
 usleep(1000);
 LCD_cmd(0x80);//Force cursor to beginning ( 1st line)
 usleep(100);
 LCD_string (" BOOK NAME: ");
 usleep(1000);
 LCD_cmd(0xC0);//Force cursor to beginning ( 2nd line)
 usleep(100);
 LCD_string (" C PROGRAMING ");
 usleep(1000);
   i2c_stop(); //Shutdown i2c connection
}

**Comment by mikeshams** [comment]
1. What issues did you or others might face with connection of an LCD Display to a BBB Wireless in terms of Hardware and software?
2. How do you resolve them?

**Comment by manharjotkaur** [comment]
1. The problems that we faced that there is no particular designated library to connect, so we need to set the gpio pins and make the lcd work as per our requirement. There were also some connection issues like for pin SCL and SDA connection. The I2C board which connects behind LCD has a one jumper which is used for on and off the lcd so initially it was left off side.
2. We took help online and tried to resolve the issue and also Derek Molloy's book helped us to resolve the issues that we faced.
3. We can change the font size, but for this program we kept it as 5*8, by setting the different values for DB7, DB6, DB5, DB4, DB3, DB2, DB1,DB0. For  DB3, two lines display, and Db4 can set the LCD in 8 bit mode and with DB2 the size of the character can be changed.
